### PR TITLE
Add unique index constraint to segmented users

### PR DESCRIPTION
**Pull request by benhalpern**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [x] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

We had a PR to add a unique index to segmented users [here](https://github.com/forem/forem/pull/19476).

We then [removed the unique constraint](https://github.com/forem/forem/pull/19544) since there were existing dupes we didn't account for and got the deployment to go through. However, this left us unable to close the loop on making this feature functional.

The approach was to delete all existing segmented users since these are re-generated and can be effemeral for all intents and purposes. That seems cleanest to me instead of a more complicated de-duping. Since the manual segmentation doesn't fully work with the unique constraint, no manual segments have been fully created.

Since these actions need to happen in sequence, a migration seems like the right approach vs a data update script, but again, feedback welcome.

I believe this fixes the issue (Definitely need a gut check in case there is something I'm missing).

## Related Tickets & Documents

- Closes https://github.com/forem/forem/issues/19576

## Added/updated tests?

- [ ] Yes
- [x] No, and this is why: Existing tests should affirm functionality.
- [ ] I need help with writing tests

### CODE REVIEW

**Review by filleduchaos — CHANGES_REQUESTED** [review]
I'm only blocking merging because the automated segments also use the `segmented_users` table; we might need to re-populate them post-deploy with the [existing data update script](https://github.com/forem/forem/blob/35a2f8b026671cf55f5d62678da8046e5ccc2110/lib/data_update_scripts/20230330152319_create_initial_audience_segments.rb) (cc @jaw6 )

Alternatively we could explicitly de-dupe the table during the migration instead of deleting all records? I know Postgres has a `DELETE USING` construct that's useful for removing duplicates - syntax would be something like

```sql
DELETE
FROM segmented_users deduped_seg_users
USING segmented_users current_seg_users
WHERE
    deduped_seg_users.id < current_seg_users.id AND
    deduped_seg_users.user_id = current_seg_users.user_id AND
    deduped_seg_users.audience_segment_id = current_seg_users.audience_segment_id;
```

which would retain only the newest records and delete older duplicates.

**Comment by benhalpern** [comment]
Yeah @filleduchaos if my approach is a little too far on the sledge hammer approach we could rewrite the deletion. Let's have that discussion.

I labeled the issue as `blocked` as well just to reinforce that this need to not move this along hastely.

_My_ understanding is that the automated segments will organically repopulate, and that technically we'll have a period of time where those segments will result in zero users, but that is okay.

But maybe I'm being silly and we should just run a de-dupe deletion query as you point out?

File: db/migrate/20230622151208_remove_existing_segmented_users.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+class RemoveExistingSegmentedUsers < ActiveRecord::Migration[7.0]
+  def up
+    SegmentedUser.delete_all
```
Comment (jaw6) [inline db/migrate/20230622151208_remove_existing_segmented_users.rb:3]: I'm fine with the de-dupe query, but just for having options: we could also tell audience segments to refresh themselves, if we don't want to wait until the overnight cron.

```suggestion
    SegmentedUser.delete_all
    AudienceSegment.find_each &:refresh!
```

**Comment by benhalpern** [comment]
Okay I went with @filleduchaos's suggestion. Seems like the fewest moving pieces at the end of the day. 👍

**Review by jaw6 — APPROVED** [review]

**Review by filleduchaos — APPROVED** [review]
🎉

**Pull request merged**